File: generated_code/python/solution_3915_3.py
from typing import List
from collections import defaultdict

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        # Step 1: Map each unique bitmask to the maximum number that produces it.
        # Since nums[i] <= 10^6, its binary representation fits within 20 bits.
        # Grouping by mask optimizes the search space significantly.
        max_for_mask = defaultdict(int)
        for num in nums:
            if num > max_for_mask[num]:
                max_for_mask[num] = num
        
        # Step 2: Extract the unique masks and sort them by value descending
        # to maximize the chance of finding larger products early.
        masks = sorted(max_for_mask.keys(), key=lambda x: max_for_mask[x], reverse=True)
        
        max_product = 0
        n = len(masks)
        
        # Step 3: Compare pairs of masks.
        # If the product of the current pair's values cannot beat the global maximum,
        # we can prune the inner loop early because the values are sorted descending.
        for i in range(n):
            val_i = max_for_mask[masks[i]]
            
            # Pruning: if squaring the current largest value cannot exceed max_product,
            # no subsequent pairs can possibly beat it.
            if val_i * val_i <= max_product:
                break
                
            for j in range(i + 1, n):
                mask_i = masks[i]
                mask_j = masks[j]
                val_j = max_for_mask[mask_j]
                
                # Pruning: if the product with the current val_j is less than or
                # equal to max_product, further elements in the j loop will be even smaller.
                if val_i * val_j <= max_product:
                    break
                
                # Check if the two numbers share no common set bits
                if (mask_i & mask_j) == 0:
                    max_product = max(max_product, val_i * val_j)
                    
        return max_product

File: generated_code/cpp/solution_3915_3.cpp
#include <vector>
#include <algorithm>
#include <unordered_map>

using namespace std;

class Solution {
public:
    long long maxProduct(vector<int>& nums) {
        // Since multiple numbers can have the same bitmask representation,
        // we only need to keep track of the maximum value for each unique bitmask.
        // This optimizes the search space significantly.
        unordered_map<int, int> max_val_for_mask;
        for (int num : nums) {
            int mask = 0;
            int temp = num;
            // Construct the bitmask representing which bits are set in the number.
            // Since nums[i] <= 10^6, it fits within 20 bits (2^20 = 1,048,576).
            for (int bit = 0; bit < 20; ++bit) {
                if ((temp >> bit) & 1) {
                    mask |= (1 << bit);
                }
            }
            max_val_for_mask[mask] = max(max_val_for_mask[mask], num);
        }

        // Convert the map to a vector of pairs for faster iteration.
        vector<pair<int, int>> mask_val_pairs;
        for (auto& p : max_val_for_mask) {
            mask_val_pairs.push_back(p);
        }

        long long max_prod = 0;
        int n = mask_val_pairs.size();

        // Compare all pairs of unique masks to find two that don't share any set bits.
        for (int i = 0; i < n; ++i) {
            int mask1 = mask_val_pairs[i].first;
            long long val1 = mask_val_pairs[i].second;
            
            for (int j = i + 1; j < n; ++j) {
                int mask2 = mask_val_pairs[j].first;
                long long val2 = mask_val_pairs[j].second;

                // If the bitwise AND is 0, they do not share any common set bits.
                if ((mask1 & mask2) == 0) {
                    max_prod = max(max_prod, val1 * val2);
                }
            }
        }

        return max_prod;
    }
};